src/main/kotlin/org/rust/lang/core/completion/RsLintCompletionProvider.kt

```diff
@@ -0,0 +1,545 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.completion
+
+import com.intellij.codeInsight.completion.CompletionParameters
+import com.intellij.codeInsight.completion.CompletionResultSet
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.patterns.ElementPattern
+import com.intellij.patterns.PlatformPatterns
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.parentOfType
+import com.intellij.util.ProcessingContext
+import org.rust.ide.icons.RsIcons
+import org.rust.ide.icons.multiple
+import org.rust.lang.RsLanguage
+import org.rust.lang.core.RsPsiPattern
+import org.rust.lang.core.psi.RsMetaItem
+import org.rust.lang.core.psi.RsPath
+import org.rust.lang.core.psi.ext.qualifier
+import org.rust.lang.core.psiElement
+import javax.swing.Icon
+
+object RsLintCompletionProvider : RsCompletionProvider() {
+    private const val RUSTC_SPECIFIC_PRIORITY = 5.0
+    private const val RUSTC_GROUP_PRIORITY = 4.0
+    private const val CLIPPY_PRIORITY = 3.0
+
+    private val GROUP_ICON = RsIcons.ATTRIBUTE.multiple()
+
+    override fun addCompletions(parameters: CompletionParameters,
+                                context: ProcessingContext,
+                                result: CompletionResultSet) {
+        val path = parameters.position.parentOfType<RsPath>() ?: return
+        val prefix = getPathPrefix(path)
+
+        LINTS.forEach {
+            val name = it.name
+            if (!name.startsWith(prefix)) return@forEach
+
+            val offered = name.removePrefix(prefix)
```

Comment:
> Right now on the root level it offers all lints (i.e. both unused and clippy::xxx), maybe we can only offer clippy:: at the root level and only after completing that we can start completing clippy lints?

It looks quite hacky. I think it's definitely better to suggest complex lints like `clippy::xxx` in two steps.
Probably, it makes sense to split this provider into two: one for `rustc` lints and another one for `clippy`